File: 2025/day12-1.hs
-- TODO: Test between best and worst case
-- TODO: Tile is just the sum of "#"
-- TODO: Region requires just the total area

import Data.Text (pack, splitOn, unpack)

type Tile = [[Bool]]

type Region = ([Int], [Int])

solution :: String -> Int
solution content = length . filter (test sizes) $ regions
  where
    (tiles, regions) = parse content
    sizes = map size tiles

test :: [Int] -> Region -> Bool
test sizes ([width, height], rs) = (width * height) >= sum (zipWith (*) sizes rs)

size :: Tile -> Int
size = sum . map (length . filter id)

parse :: String -> ([Tile], [Region])
parse content = (map parseTile . init $ parts, map parseRegion . lines . last $ parts)
  where
    parts = map unpack . splitOn (pack "\n\n") . pack $ content

parseTile :: String -> Tile
parseTile = map (map (== '#')) . drop 1 . lines

parseRegion :: String -> Region
parseRegion line = (size, map read . drop 1 $ parts)
  where
    parts = words line
    size = map (read . unpack) . splitOn (pack "x") . pack . init . (!! 0) $ parts

main :: IO ()
main = do readFile "day12.dat" >>= print . solution

File: 2025/day12-1.ts
// TODO: Test between best and worst case
// TODO: Region has a width and height

type Region = {
  area: number;
  requests: number[];
};

function solution(content: string) {
  const parts = content.split("\n\n");
  const tiles = parts.slice(0, -1).map(parseTile);
  const regions = parts.at(-1)!.split("\n").map(parseRegion);
  return regions.filter((region) => test(tiles, region)).length;
}

function test(sizes: number[], region: Region): boolean {
  const { area, requests } = region;
  return (
    area >= requests.reduce((acc, cur, index) => acc + cur * sizes[index], 0)
  );
}

function parseTile(part: string): number {
  return part.split("").filter((item) => item === "#").length;
}

function parseRegion(part: string): Region {
  const matches = part.match(/(\d+)x(\d+): (.+)/);
  if (matches === null) throw new Error(`Invalid line: ${part}`);
  return {
    area: Number(matches[1]) * Number(matches[2]),
    requests: matches[3].split(" ").map(Number),
  };
}

Deno.readTextFile("day12.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
